apis/python/src/tiledbsoma/_read_iters.py

```diff
@@ -5,59 +5,424 @@
 
 """Read iterators.
 """
+from __future__ import annotations
+
 import abc
-from typing import TypeVar
+from concurrent.futures import ThreadPoolExecutor
+from typing import (
+    TYPE_CHECKING,
+    Dict,
+    Iterator,
+    List,
+    Optional,
+    Sequence,
+    Tuple,
+    TypeVar,
+    Union,
+    cast,
+)
 
 import numpy as np
+import numpy.typing as npt
+import pandas as pd
 import pyarrow as pa
 import somacore
+from scipy import sparse
+from somacore import options
+from somacore.query._eager_iter import EagerIterator
 
 # This package's pybind11 code
 import tiledbsoma.pytiledbsoma as clib
 
+from . import _util
+from ._exception import SOMAError
 from ._types import NTuple
 
+if TYPE_CHECKING:
+    from . import SparseNDArray
+
+
+# Convenience types
+_RT = TypeVar("_RT")
+BlockwiseTableReadIterResult = Tuple[pa.Table, Tuple[pa.Array, ...]]
+BlockwiseSingleAxisTableIter = Iterator[BlockwiseTableReadIterResult]
+
+BlockwiseScipyReadIterResult = Tuple[
+    Union[sparse.csr_matrix, sparse.csc_matrix, sparse.coo_matrix],
+    Tuple[npt.NDArray[np.int64], npt.NDArray[np.int64]],
+]
+
+IndicesType = Tuple[npt.NDArray[np.int64], npt.NDArray[np.int64]]
+IJDType = Tuple[
+    Tuple[npt.NDArray[np.int64], npt.NDArray[np.int64]], npt.NDArray[np.generic]
+]
+
 
 class TableReadIter(somacore.ReadIter[pa.Table]):
     """Iterator over `Arrow Table <https://arrow.apache.org/docs/python/generated/pyarrow.Table.html>`_ elements"""
 
     def __init__(self, sr: clib.SOMAArray):
-        self.sr = sr
+        self._reader = _arrow_table_reader(sr)
 
     def __next__(self) -> pa.Table:
-        arrow_table = self.sr.read_next()
-        if arrow_table is None:
-            raise StopIteration
-
-        return arrow_table
+        return next(self._reader)
 
     def concat(self) -> pa.Table:
         """Concatenate remainder of iterator, and return as a single `Arrow Table <https://arrow.apache.org/docs/python/generated/pyarrow.Table.html>`_"""
         return pa.concat_tables(self)
 
 
-RT = TypeVar("RT")
+class BlockwiseReadIterBase(somacore.ReadIter[_RT], metaclass=abc.ABCMeta):
+    """Private implementation class. Currently implemented as a single-axis blockwise iterator"""
+
+    _reader: Iterator[_RT]
+
+    def __init__(
+        self,
+        array: "SparseNDArray",
+        sr: clib.SOMAArray,
+        coords: options.SparseNDCoords,
+        axis: Union[int, Sequence[int]],
+        *,
+        size: Optional[Union[int, Sequence[int]]] = None,
+        reindex_disable: Optional[Union[int, Sequence[int]]] = None,
+        eager: bool = True,
+        pool: Optional[ThreadPoolExecutor] = None,
+    ):
+        super().__init__()
+
+        self.ndim = len(sr.shape)
+        self.array = array
+        self.sr = sr
+        self.eager = eager
+        self.pool = pool
```

Comment:
Would you want to make these `_`-prefixed to make it clear that they should not be modified by outsiders? Up to you; the fact that they are in an internal class that a user will never have cause to instantiate might be enough.

Reply:
I had assumed that it was OK as it was an implementation class.  But I can do it if it help

Are you suggest it for the instance fields, or the arguments to \_\_init__?

Reply:
Just the instance fields, but as I said, this isn’t critical.